rfcs/20210301-computation-datatype/README.md

```diff
@@ -0,0 +1,178 @@
+  typically interchangeable for training.
+
+Proposal: support only down-conversion to "compatible"
+data-types. This will allow to restrict the feature to only commonly
+used combinations and would make internal logic and validation
+simpler. Here we would return `invalid_argument` if users pass
+incompatible data-type for down-conversion hint/requirement, to help
+them debug incorrect logic.
+
+## Down-conversion policy: requirement vs hint.
+Here the question is if the user specifies a target type to
+down-convert to, should we treat it as a requirement (we can
+down-convert only to this type) or should it be a hint (we can
+down-convert to a type other than the one specified as long as it is
+"compatible").
+
+Proposal: support only hints, to allow library for optimal dispatching
+without user needing to tweak down-conversion policy for better
+performance. Also, this is more scalable as we introduce new
+data-types (e.g. if we were to introduce `tf32`).
+
+## Per tensor vs per primitive setting
+This would be the granularity at which the user would allow conversion
+to happen. Per tensor seems to have the most flexibility, however it
+is unlikely that users would need this level of flexibility. For the
+per-primitive setting, the implicit conversion will be assumed to
+happen only for input tensors, and not output.
+
+Proposal: support only per primitive setting. The main points against
+per tensor are: a) it would require more user changes to enable this
+feature and b) the implicit down-convert feature is planned to be
+exposed to end-user (FWK or toolkit user), and there is no plan to
+expose any fine grain control there (it will be global or scoped to a
+sub-graph).
+
+## Allow down-conversion through reorder vs hide it in execution
+
+This is basically deciding what data-type should be returned when a
+primitive is queried for one of its memory descriptor when implicit
+down-conversion happens. One option is to preserve the data-type
+passed by user so that down-conversion is fully hidden and
+implicit. The other option is to return a memory descriptor with the
+data-type that will be used during computation. The first option has
+the benefit to completely hide the down-conversion and does not
+require the user to add logic to handle potential down-conversion
+through reorder if creation data-type is different than queried
+data-type. The second option has the benefit of allowing to
+down-convert ahead of time and reduce the execution time (e.g. weights
+down-conversion can be computed ahead of time during inference).
+
+Proposal: expose the data-type we down-convert to only when the user
+data layout is `any`. When the user does not pass `any`, we cannot
+assume that they properly handle a potential reorder whereas when they
+pass `any`, we can assume that a memory_descriptor comparison happens
+(using `dnnl_memory_desc_equal` or `memory::desc::operator==`) and
+proper reorder logic is implemented.
```

Comment:
One potential issue with that is increased memory consumption and extra reorders. With `any` and enabled down-conversion this approach forces users to a create temporary buffer and reorder to it.

With the alternative approach (no implicit data type change for `any`) we can do that on the fly, reading `f32` and making conversion on registers (this is optional, such behavior would be up to the implementation). With that we can save one reorder and one extra tensor buffer. I guess in some cases (for small sizes?) the approach without reorders would be more performant. But overall I'm not sure how much of an issue it is.

Reply:
When a user creates a primitive with one of the layouts being `any`, a reorder can happen, regardless of down-conversion.
If you are mostly concerned about activations, where in some cases we preserve plain layouts (even when use pass `any`) to minimize potential reorders overheads, nothing prevents us to do the same with this proposals. In those cases, we would still return plain layout and the user datatype to prevent reorders.

Here the main intent when allowing to expose the datatypes we down-convert to is for weights during inference.
The hidden conversion option could incur overhead when weights are much bigger than activations, and not much data reuse happen (think GEMM for small batch inference for example). In those cases:
- pre-computing the down-conversion should help save bandwidth (up to ~2x for GEMV cases)
- pre-computing the down-conversion would reduce the kernel launch overhead (for GEMM, separate kernels are launched for packing and computing).

Reply:
I was mostly thinking about stacked convolutions. If we decide to expose data type down-conversion via reorders it would mean we end up with something like:

- ...
- Reorder for source `f32` -> `f16`  (layout is preserved, assume weights reorder is done beforehand)
- Convolution: `f16` -> `f32` (destination type is `f32`, as the user requested)
- Reorder for source `f32` -> `f16` (layout is preserved, only down-conversion)
- Convolution: `f16` -> `f32`
- ...

The other approach assumes no reorders with down-conversion is happening inside the kernel.

I think it would be fine to keep this as an option and let the implementation decide if it's better to expose down-conversion in case of `any` or keep it inside the kernel. In this case we say to users: with `any`, they may have both 1) layout change and 2) data type change but it's implementation-dependent.

Reply:
Seems you already mentioned that this will not be a strict rule to follow:

>.. to minimize potential reorders overheads, nothing prevents us to do the same with this proposals. In those cases, we would still return plain layout and the user datatype to prevent reorders.

In the RFC, I was slightly confused by this (as initially I read this as a mandatory rule to follow):

> Proposal: expose the data-type we down-convert to only when the user data layout is any.

Reply:
Ok, my bad, I will clarify this point.